src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/judgeapproval/JudgeApprovalResolver.java

```diff
@@ -0,0 +1,81 @@
+import uk.gov.hmcts.reform.finrem.caseorchestration.model.ccd.wrapper.DraftOrdersWrapper;
+import uk.gov.hmcts.reform.finrem.caseorchestration.service.IdamService;
+
+import java.time.LocalDateTime;
+import java.util.Arrays;
+import java.util.List;
+
+import static java.util.Optional.ofNullable;
+import static uk.gov.hmcts.reform.finrem.caseorchestration.model.ccd.draftorders.judgeapproval.JudgeDecision.JUDGE_NEEDS_TO_MAKE_CHANGES;
+import static uk.gov.hmcts.reform.finrem.caseorchestration.model.ccd.draftorders.judgeapproval.JudgeDecision.READY_TO_BE_SEALED;
+
+@Component
+@RequiredArgsConstructor
+class JudgeApprovalResolver {
+
+    private final IdamService idamService;
+
+    private final HearingProcessor hearingProcessor;
+
+    void populateJudgeDecision(DraftOrdersWrapper draftOrdersWrapper, CaseDocument targetDoc, JudgeApproval judgeApproval,
+                                         String userAuthorisation) {
+        ofNullable(draftOrdersWrapper.getDraftOrdersReviewCollection())
+            .ifPresent(collection -> processApprovableCollection(collection.stream()
+                .flatMap(c -> c.getValue().getDraftOrderDocReviewCollection().stream().map(DraftOrderDocReviewCollection::getValue))
+                .toList(), targetDoc, judgeApproval, userAuthorisation));
+
+        ofNullable(draftOrdersWrapper.getDraftOrdersReviewCollection())
+            .ifPresent(collection -> processApprovableCollection(collection.stream()
+                .flatMap(c -> c.getValue().getPsaDocReviewCollection().stream().map(PsaDocReviewCollection::getValue))
+                .toList(), targetDoc, judgeApproval, userAuthorisation));
+
+        ofNullable(draftOrdersWrapper.getAgreedDraftOrderCollection())
+            .ifPresent(agreedDraftOrderCollections ->
+                processApprovableCollection(agreedDraftOrderCollections.stream().map(AgreedDraftOrderCollection::getValue).toList(), targetDoc,
+                    judgeApproval, userAuthorisation));
+
+        if (isJudgeApproved(judgeApproval)) {
+            ofNullable(draftOrdersWrapper.getHearingInstruction())
+                .map(HearingInstruction::getAnotherHearingRequestCollection)
+                .ifPresent(collection -> collection.forEach(a -> hearingProcessor.processHearingInstruction(draftOrdersWrapper, a.getValue())));
+        }
+    }
+
+    void processApprovableCollection(List<? extends Approvable> approvables, CaseDocument targetDoc, JudgeApproval judgeApproval,
+                                               String userAuthorisation) {
+        ofNullable(approvables)
+            .ifPresent(list ->
+                list.forEach(el -> ofNullable(el)
+                    .filter(approvable -> approvable.match(targetDoc))
+                    .ifPresent(approvable -> handleApprovable(approvable, judgeApproval, userAuthorisation))
+                )
+            );
+    }
```

Comment:
Javadoc for the above method would be good.

Reply:
[e032596](https://github.com/hmcts/finrem-case-orchestration-service/pull/1997/commits/e032596adbc8224bf8b416ece1089447a1bda715)